### PR TITLE
New settings feature switch and fix for more missing and more than one channel

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.11.3
- Added: New settings to allow feature switches for generating repeating timers and autotimers
- Added: Config guide to Readme
- Fixed: Missing channel or more than one channel in timers/autotimers

### CODE REVIEW

**Comment by phunkyfish** [comment]
I have those changes done. Let me test once I’m home from work and I’ll send you a note once I have it pushed.

**Comment by phunkyfish** [comment]
Ok, we should be good.

File: src/client.cpp [hunk 1]
```diff
@@ -269,260 +269,99 @@ void ADDON_Destroy()
   m_CurStatus = ADDON_STATUS_UNKNOWN;
 }
 
-ADDON_STATUS ADDON_SetSetting(const char *settingName, const void *settingValue)
+ADDON_STATUS SetStringSetting(const string settingName, const char* settingValue, string currentValue, ADDON_STATUS returnValueIfChanged)
```
Comment (ksooo) [inline src/client.cpp]: You should always keep the `std::` namespace. This improves readability and is common sense. `string` -> `std::string`.

And prefer by-reference parameters over by-value. `string`-> `const std::string&`

`currentValue`seems to be an inout param. Thus it must not be a by-value parameter!

Corrected signature: `ADDON_STATUS SetStringSetting(const std::string& settingName, const char* settingValue,  std::string& currentValue, ADDON_STATUS returnValueIfChanged)`

File: src/client.cpp [hunk 2]
```diff
@@ -269,260 +269,99 @@ void ADDON_Destroy()
...
+  if (strSettingValue != currentValue)
   {
-    string tmp_sHostname;
-    XBMC->Log(LOG_INFO, "%s - Changed Setting 'host' from %s to %s", __FUNCTION__, g_strHostname.c_str(), (const char*) settingValue);
-    tmp_sHostname = g_strHostname;
-    g_strHostname = (const char*) settingValue;
-    if (tmp_sHostname != g_strHostname)
-      return ADDON_STATUS_NEED_RESTART;
+    XBMC->Log(LOG_INFO, "%s - Changed Setting '%s' from %s to %s", __FUNCTION__, settingName.c_str(), currentValue.c_str(), strSettingValue.c_str());
+    currentValue = strSettingValue;
```
Comment (ksooo) [inline src/client.cpp]: As `currentValue` is a by-value parameter this modifies only your local copy! The parameter type has to changed. See my other comment with the corrected signature.

File: src/client.cpp [hunk 3]
```diff
@@ -269,260 +269,99 @@ void ADDON_Destroy()
...
+    return returnValueIfChanged;
   }
-  else if (str == "webport")
+
+  return ADDON_STATUS_OK;
+}
+
+ADDON_STATUS SetIntSetting(const string settingName, int settingValue, int &currentValue, ADDON_STATUS returnValueIfChanged)
```
Comment (ksooo) [inline src/client.cpp]: => `ADDON_STATUS SetIntSetting(const std::string& settingName, int settingValue, int &currentValue, ADDON_STATUS returnValueIfChanged)`

File: src/client.cpp [hunk 4]
```diff
@@ -269,260 +269,99 @@ void ADDON_Destroy()
...
+    XBMC->Log(LOG_INFO, "%s - Changed Setting '%s' from %d to %d", __FUNCTION__, settingName.c_str(), currentValue, settingValue);
+    currentValue = settingValue;
+    return returnValueIfChanged;
   }
-  else if (str == "use_secure")
+
+  return ADDON_STATUS_OK;
+}
+
+ADDON_STATUS SetBoolSetting(const string settingName, bool settingValue, bool &currentValue, ADDON_STATUS returnValueIfChanged)
```
Comment (ksooo) [inline src/client.cpp]: => `ADDON_STATUS SetBoolSetting(const std::string& settingName, bool settingValue, bool &currentValue, ADDON_STATUS returnValueIfChanged)`

Comment (ksooo) [reply src/client.cpp]: Looking at the bodies of the three new functions, which are very similar, how about a template function?

```C++
namespace
{
template <typename T>
ADDON_STATUS SetSetting(const std::string& settingName, T settingValue, T& currentValue, ADDON_STATUS returnValueIfChanged)
{
  if (settingValue != currentValue)
  {
     XBMC->Log(LOG_INFO, "%s - Changed Setting '%s'", __FUNCTION__, settingName.c_str());
     currentValue = settingValue;
     return returnValueIfChanged;
  }
  return ADDON_STATUS_OK;
 }
} // unnamed namespace
```

Note that I have removed the logging of old and new value which would only be possible if you add a parameter for passing the right character for the `printf`of the value (which I consider ugly).

But again, this is just an idea idea. No need to follow my brain fart.

Comment (phunkyfish) [reply src/client.cpp]: Or use to_string to pass the values???

Then it’s completely the same. I’ll give it a try in the morning

**Comment by phunkyfish** [comment]
As I made a mess of the commits I’ll just reset HEAD and start over. Do you want the minor refactoring and fix in Timer.cpp/h in separate commits or are you ok with them together in one?

**Comment by ksooo** [comment]
If feasible, make separate commits, please. Makes it much easier to understand and review the changes.

File: src/client.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/client.cpp]: I had a look at this but doesn't work with extern C linkage. When I get around to a proper refactor I'll try it (Settings in own class etc).

**Comment by phunkyfish** [comment]
Ok, maybe, hopefully, this is ready now!

**Comment by phunkyfish** [comment]
And thanks again for all your guidance. I’m learning loads!

**Comment by ksooo** [comment]
> This branch has conflicts that must be resolved

Please rebase.

**Comment by phunkyfish** [comment]
Ok, rebased. Should be good now.

**Pull request merged**